### PR TITLE
Add Failure API tests for user defined federated authenticator management

### PR DESCRIPTION
**Pull request by Shenali-SJ**
### Proposed changes in this pull request

This PR adds API tests to validate failure scenarios for IdPs with user-defined authenticators.

**Related Issue:**
- https://github.com/wso2/product-is/issues/21216

**Related PR:**
- [Draft] https://github.com/wso2/product-is/pull/21646

---

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPSuccessTest.java [hunk 1]
```diff
@@ -61,12 +79,37 @@ public IdPSuccessTest(TestUserMode userMode) throws Exception {
...
+        userDefinedAuthenticatorPayload.setEndpoint(endpoint);
+
+        return userDefinedAuthenticatorPayload;
     }
 
     @AfterClass(alwaysRun = true)
     public void testConclude() {
 
         super.conclude();
+        customIdPId = null;
```
Comment (Thisara-Welmilla) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPSuccessTest.java:112]: Do we really need this?

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java [hunk 2]
```diff
@@ -76,15 +105,7 @@ public void testInit() {
     public void testFinish() {
 
         RestAssured.basePath = StringUtils.EMPTY;
-    }
-
-    @DataProvider(name = "restAPIUserConfigProvider")
```
Comment (Thisara-Welmilla) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java:81]: any reason to remove this?

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java [hunk 3]
```diff
@@ -179,4 +199,248 @@ public void testPatchIdPNonExistentProperties() throws IOException {
...
+                .assertThat()
+                .statusCode(HttpStatus.SC_CREATED)
+                .header(HttpHeaders.LOCATION, notNullValue());
+        String location = response.getHeader(HttpHeaders.LOCATION);
+        assertNotNull(location);
+        customIdPId = location.substring(location.lastIndexOf("/") + 1);
+
+        // duplicate the authenticator creation
+        duplicatedUserDefinedAuthenticatorPayload =
+                createUserDefinedAuthenticatorPayload("Y3VzdG9tQXV0aGVudGljYXRvcjE=",
```
Comment (Thisara-Welmilla) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java:327]: Lets make this string constants as we are using them mutiple places.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPSuccessTest.java (hunk 1 above)
Comment (Shenali-SJ) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPSuccessTest.java]: This is removed with d6716dba397569d02ce08f1776ca188a75c02680

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java (hunk 3 above)
Comment (Shenali-SJ) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java]: This is updated with d6716dba397569d02ce08f1776ca188a75c02680

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java [hunk 4]
```diff
@@ -76,15 +108,7 @@ public void testInit() {
...
-    }
-
-    @DataProvider(name = "restAPIUserConfigProvider")
-    public static Object[][] restAPIUserConfigProvider() {
-
-        return new Object[][]{
-                {TestUserMode.SUPER_TENANT_ADMIN},
-                {TestUserMode.TENANT_ADMIN}
-        };
+        customIdPId = null;
```
Comment (Thisara-Welmilla) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java:111]: This does not delete the idp from the IS. Delete the idp here.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java [hunk 5]
```diff
@@ -179,4 +202,244 @@ public void testPatchIdPNonExistentProperties() throws IOException {
...
+//                .header(HttpHeaders.LOCATION, notNullValue());
+//
+//        String location = response.getHeader(HttpHeaders.LOCATION);
+//        assertNotNull(location);
+//        String customIdPId = location.substring(location.lastIndexOf("/") + 1);
+//        assertNotNull(customIdPId);
+//    }
+
+    @Test
+    public void testAddIdPWithUserDefinedAuthenticatorWhenAuthenticatorNameIsDuplicated() throws IOException {
```
Comment (Thisara-Welmilla) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java:311]: testAddIdPWithUserDefinedAuthenticatorWhenAuthenticatorNameIsDuplicated -> testAddIdPWithUserDefinedAuthenticatorWithExistingAuthenticatorName

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java [hunk 6]
```diff
@@ -179,4 +202,244 @@ public void testPatchIdPNonExistentProperties() throws IOException {
...
+
+        String idpCreateErrorPayload = readResource("add-idp-with-custom-fed-multi-auth.json");
+
+        UserDefinedAuthenticatorPayload userDefinedAuthenticatorPayload1 = createUserDefinedAuthenticatorPayload(
+                "Y3VzdG9tQXV0aGVudGljYXRvclg=",
+                "https://abc.com/authenticate",
+                "testUser",
+                "testPassword");
+        UserDefinedAuthenticatorPayload userDefinedAuthenticatorPayload2 = createUserDefinedAuthenticatorPayload(
+                "Y3VzdG9tQXV0aGVudGljYXRvclg=",
```
Comment (Thisara-Welmilla) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java:355]: Use deferent authenticator ids

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java (hunk 4 above)
Comment (Shenali-SJ) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java]: This is fixed.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java (hunk 6 above)
Comment (Shenali-SJ) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java]: This is updated with 204aaea3e005918dd0879436e47ffd04e39ad2a8

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java [hunk 7]
```diff
@@ -78,15 +108,6 @@ public void testFinish() {
...
 
-    @DataProvider(name = "restAPIUserConfigProvider")
-    public static Object[][] restAPIUserConfigProvider() {
-
-        return new Object[][]{
-                {TestUserMode.SUPER_TENANT_ADMIN},
-                {TestUserMode.TENANT_ADMIN}
-        };
-    }
-
```
Comment (Thisara-Welmilla) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java:89]: Is there any reason why we bring it up.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java [hunk 8]
```diff
@@ -179,4 +199,260 @@ public void testPatchIdPNonExistentProperties() throws IOException {
...
+
+        Response response = getResponseOfPostNoFilter(IDP_API_BASE_PATH, body);
+        response.then()
+                .log().ifValidationFails()
+                .assertThat()
+                .statusCode(HttpStatus.SC_BAD_REQUEST)
+                .body("message", equalTo("Endpoint configuration must be provided for the user defined " +
+                        "federated authenticators " + new String(Base64.getDecoder().decode(AUTHENTICATOR_ID_1)) + "."));
+    }
+
```
Comment (Thisara-Welmilla) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java:265]: Let's add test when one of authenticator properties are missing. eg: username in BASIC

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java [hunk 9]
```diff
@@ -179,4 +199,260 @@ public void testPatchIdPNonExistentProperties() throws IOException {
...
+                        "testUser1",
+                        "testPassword1");
+        Response responseOfDuplicate = createUserDefAuthenticator("CustomAuthIDP2",
+                duplicatedUserDefinedAuthenticatorPayload);
+        responseOfDuplicate.then()
+                .log().ifValidationFails()
+                .assertThat()
+                .statusCode(HttpStatus.SC_BAD_REQUEST)
+                .body("message", equalTo("Federated authenticator name " +
+                        new String(Base64.getDecoder().decode(AUTHENTICATOR_ID_1)) + " is already taken."));
```
Comment (Thisara-Welmilla) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java:314]: Delete the created idp in this method

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java [hunk 10]
```diff
@@ -179,4 +199,260 @@ public void testPatchIdPNonExistentProperties() throws IOException {
...
+        Response response = getResponseOfPostNoFilter(IDP_API_BASE_PATH, body);
+        response.then()
+                .log().ifValidationFails()
+                .assertThat()
+                .statusCode(HttpStatus.SC_BAD_REQUEST)
+                .body("message", equalTo("Endpoint configuration must be provided for the user defined " +
+                        "federated authenticators " + new String(Base64.getDecoder().decode(AUTHENTICATOR_ID_1)) + "."));
+    }
+
+    @Test
```
Comment (Thisara-Welmilla) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java:266]: Add a test where adding custom authenticator with system defined authenticator id.

Comment (Thisara-Welmilla) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java:266]: Add a test where given custom authenticator ID with SYSTEM definedby prop.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java (hunk 7 above)
Comment (Shenali-SJ) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java]: This is fixed with 53c5cf7c40298253d04e691e9222d21f771638fe

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java (hunk 10 above)
Comment (Shenali-SJ) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java:253]: This is updated with 104bc9c032e465bb70bd9f8c1f0dd364ba62aa37

Comment (Shenali-SJ) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java:253]: This is updated with 104bc9c032e465bb70bd9f8c1f0dd364ba62aa37

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java (hunk 9 above)
Comment (Shenali-SJ) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java]: This is updated with 104bc9c032e465bb70bd9f8c1f0dd364ba62aa37

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java (hunk 8 above)
Comment (Shenali-SJ) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/IdPFailureTest.java:252]: This is updated with 104bc9c032e465bb70bd9f8c1f0dd364ba62aa37

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/common/RESTTestBase.java [hunk 11]
```diff
@@ -380,6 +380,18 @@ protected Response getResponseOfPost(String endpointUri, String body) {
                 .post(endpointUri);
     }
 
+    protected Response getResponseOfPostNoFilter(String endpointUri, String body) {
```
Comment (Thisara-Welmilla) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/common/RESTTestBase.java:383]: Add a method comment

Comment (Shenali-SJ) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/common/RESTTestBase.java:392]: This is updated with e24b571fe77c0a70265d04002039fee854912c9e

**Review by Thisara-Welmilla — APPROVED** [review]

**Review by Thisara-Welmilla — APPROVED** [review]

**Review by Thisara-Welmilla — APPROVED** [review]

**Pull request merged**